crypto/store/loader_file.c

```diff
@@ -0,0 +1,1025 @@
+                                                const unsigned char *blob,
+                                                size_t len, void **handler_ctx,
+                                                const UI_METHOD *ui_method,
+                                                void *ui_data);
+/*
+ * The eof function should return 1 if there's no more data to be found
+ * with the handler_ctx, otherwise 0.  This is only used when the handler is
+ * marked repeatable.
+ */
+typedef int (*STORE_FILE_eof_fn)(void *handler_ctx);
+/*
+ * The destroy_ctx function is used to destroy the handler_ctx that was
+ * intiated by a repeatable try_decode fuction.  This is only used when
+ * the handler is marked repeatable.
+ */
+typedef void (*STORE_FILE_destroy_ctx_fn)(void **handler_ctx);
+
+typedef struct store_file_handler_st {
+    const char *name;
+    STORE_FILE_try_decode_fn try_decode;
+    STORE_FILE_eof_fn eof;
+    STORE_FILE_destroy_ctx_fn destroy_ctx;
+
+    /* flags */
+    int repeatable;
+} STORE_FILE_HANDLER;
+
+static STORE_INFO *try_decode_PKCS12(const char *pem_name,
+                                     const char *pem_header,
+                                     const unsigned char *blob, size_t len,
+                                     void **pctx, const UI_METHOD *ui_method,
+                                     void *ui_data)
+{
+    STORE_INFO *store_info = NULL;
+    STACK_OF(STORE_INFO) *ctx = *pctx;
+
+    if (ctx == NULL) {
+        /* Initial parsing */
+        PKCS12 *p12;
+        int ok = 0;
+
+        if (pem_name != NULL)
+            /* No match, there is no PEM PKCS12 tag */
+            return NULL;
+
+        if ((p12 = d2i_PKCS12(NULL, &blob, len)) != NULL) {
+            char *pass = NULL;
+            char tpass[PEM_BUFSIZE];
+            EVP_PKEY *pkey = NULL;
+            X509 *cert = NULL;
+            STACK_OF(X509) *chain = NULL;
+
+            if (PKCS12_verify_mac(p12, "", 0)
+                || PKCS12_verify_mac(p12, NULL, 0)) {
+                pass = "";
+            } else {
+                if ((pass = file_get_pass(ui_method, tpass, PEM_BUFSIZE,
+                                          "PKCS12 import password",
+                                          ui_data)) == NULL) {
+                    STOREerr(STORE_F_TRY_DECODE_PKCS12,
+                             STORE_R_PASSPHRASE_CALLBACK_ERROR);
+                    goto p12_end;
+                }
+                if (!PKCS12_verify_mac(p12, pass, strlen(pass))) {
+                    STOREerr(STORE_F_TRY_DECODE_PKCS12,
+                             STORE_R_ERROR_VERIFYING_PKCS12_MAC);
+                    goto p12_end;
+                }
+            }
+
+            if (PKCS12_parse(p12, pass, &pkey, &cert, &chain)) {
+                if ((ctx = sk_STORE_INFO_new_null()) != NULL
+                    && sk_STORE_INFO_push(ctx, STORE_INFO_new_PKEY(pkey)) != 0
+                    && sk_STORE_INFO_push(ctx, STORE_INFO_new_CERT(cert)) != 0) {
```

Comment:
`ok` will be zero, `ctx` will be freed and set to NULL, and `store_info` will be NULL as a result.